### PR TITLE
Do not write output key whose value is identical to input

### PR DESCRIPTION
**Pull request by miklos-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala [hunk 1]
```diff
@@ -39,12 +42,24 @@ private[kvutils] trait CommitContext {
 
   /** Set a value in the output state. */
   def set(key: DamlStateKey, value: DamlStateValue): Unit = {
-    if (!outputs.contains(key)) {
-      outputOrder += key
+    if (inputAlreadyContains(key, value)) {
+      logger.error(s"Identical output found for key $key")
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:46]: Why an error?

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:46]: Will change that before merging -- wanted to see if we actually hit this.

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Changed it back to `TRACE`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala [hunk 2]
```diff
@@ -59,4 +62,11 @@ private[kvutils] trait CommitContext {
   def getOutputs: Iterable[(DamlStateKey, DamlStateValue)] =
     outputOrder.map(k => k -> outputs(k))
 
+  private def inputAlreadyContains(key: DamlStateKey, value: DamlStateValue): Boolean =
+    inputs
+      .get(key)
+      .exists {
+        case Some(inputValue) => inputValue == value
+        case None => false
+      }
```
Comment (andreaslochbihler-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:71]: Shorter:
```suggestion
    inputs.get(key).exists(_.contains(value))
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala [hunk 3]
```diff
@@ -8,16 +8,19 @@ import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
...
 
 /** Commit context provides access to state inputs, commit parameters (e.g. record time) and
   * allows committer to set state outputs.
   */
 private[kvutils] trait CommitContext {
+  private[this] val logger = LoggerFactory.getLogger(this.getClass)
+
   def inputs: DamlStateMap
   // NOTE(JM): The outputs must be iterable in deterministic order, hence we
   // keep track of insertion order.
```
Comment (andreaslochbihler-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:26]: Every `Committer` uses its own `CommitContext`, so the output order is determinized for each DamlSubmission in isolation. The output order across several DamlSubmissions does not seem to be deterministic given that `BatchedSubmissionValidator` calls `commit`s inside a `mapAsync`. I'm not sure though why we even need deterministic output order here. Can you check why we need the deterministic output order in the first place?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala [hunk 4]
```diff
@@ -39,16 +42,16 @@ private[kvutils] trait CommitContext {
 
   /** Set a value in the output state. */
   def set(key: DamlStateKey, value: DamlStateValue): Unit = {
-    if (!outputs.contains(key)) {
-      outputOrder += key
+    if (inputAlreadyContains(key, value)) {
+      logger.trace(s"Identical output found for key $key")
```
Comment (andreaslochbihler-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:46]: I'd additionally remove the `key` from `outputs` and possibly `outputOrder`. Otherwise `set` has a weird semantics. For example,
```
inputs = Map(k1 -> Some(v1))

set(k1, v2)
set(k1, v1)
```
where `v1 != v2`. When we call `getOutputs`, we will get `Seq(k1 -> v2)` even though the second `set` should have overwritten the first. I would personally find it simpler if `set` remains unchanged and we only filter the outputs in `getOutputs`.

Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:46]: If you use an Slf4j logger, you may want to use Slf4j templates instead of Scala string interpolation.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala (hunk 3 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:26]: We ensure determinisim in `BatchedSubmissionValidator` by doing the conflict detection first (according to a fixed order) and only `commit`ing submissions that won't conflict and will generate the same write set regardless of order of execution.
We need deterministic output for historical reasons here -- we may want to relax this requirement.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala (hunk 4 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:46]: Thanks, that's a good point especially in case of expensive `.toString` methods.

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala]: Applied your suggestion (i.e., doing the filtration only in `.getOutputs`) and added a test case for verifying this corner case.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala (hunk 3 above)
Comment (andreaslochbihler-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:26]: Makes sense. Thanks for explaining the historical reasons.
